### ISSUE TITLE
Concurrent Requests Lock Client

### ISSUE THREAD

**Issue by blowmage** [googleapis/gax-ruby#79]
Hello, we are unable to make more than 4 concurrent requests using a GAPIC client. When more than 4 concurrent requests are made the client seems to lock up until the requests time out. We are currently in an effort to add more concurrency in the Google Cloud Pub/Sub Ruby library, and while this seems to be working well while making concurrent streaming API calls, we are seeing the grpc client lock up while concurrently calling non-streaming APIs under moderate load.

To illustrate, I have included a ruby script that attempts to benchmark the performance differences between making sequential and concurrent requests. The default number of requests is 4, but you can pass in an argument to change the number of requests. This script will lock up and eventually time out when passing in a number larger than 4. Here is the output when running 4 and 6 concurrent requests:

```
$ ruby benchmark_gapic.rb 4
       user     system      total        real
sequential#<Google::Pubsub::V1::PublishResponse:0x007faed3afb160>
#<Google::Pubsub::V1::PublishResponse:0x007faed3ad8660>
#<Google::Pubsub::V1::PublishResponse:0x007faed3aa9bd0>
#<Google::Pubsub::V1::PublishResponse:0x007faed3a7b118>
  0.390000   0.500000   0.890000 ( 16.913990)
concurrent#<Google::Pubsub::V1::PublishResponse:0x007faed1845b20>
#<Google::Pubsub::V1::PublishResponse:0x007faed18a7898>
#<Google::Pubsub::V1::PublishResponse:0x007faed19372b8>
#<Google::Pubsub::V1::PublishResponse:0x007faed1955178>
  0.390000   0.540000   0.930000 ( 11.678385)

$ ruby benchmark_gapic.rb 6
       user     system      total        real
sequential#<Google::Pubsub::V1::PublishResponse:0x007ffd71951f48>
#<Google::Pubsub::V1::PublishResponse:0x007ffd71923418>
#<Google::Pubsub::V1::PublishResponse:0x007ffd719009b8>
#<Google::Pubsub::V1::PublishResponse:0x007ffd718d1f28>
#<Google::Pubsub::V1::PublishResponse:0x007ffd6f847488>
#<Google::Pubsub::V1::PublishResponse:0x007ffd6f8979d8>
  0.570000   0.700000   1.270000 ( 18.899312)
concurrent/Users/blowmage/.gem/repos/gcloud-ruby/gems/google-gax-0.8.4/lib/google/gax/api_callable.rb:357:in `rescue in block (2 levels) in retryable': GaxError Retry total timeout exceeded with exception, caused by 4:Deadline Exceeded (Google::Gax::RetryError)
	from /Users/blowmage/.gem/repos/gcloud-ruby/gems/google-gax-0.8.4/lib/google/gax/api_callable.rb:343:in `block (2 levels) in retryable'
	from /Users/blowmage/.gem/repos/gcloud-ruby/gems/google-gax-0.8.4/lib/google/gax/api_callable.rb:342:in `loop'
	from /Users/blowmage/.gem/repos/gcloud-ruby/gems/google-gax-0.8.4/lib/google/gax/api_callable.rb:342:in `block in retryable'
	from /Users/blowmage/.gem/repos/gcloud-ruby/gems/google-gax-0.8.4/lib/google/gax/api_callable.rb:262:in `block in catch_errors'
	from /Users/blowmage/.gem/repos/gcloud-ruby/gems/google-gax-0.8.4/lib/google/gax/api_callable.rb:224:in `block in create_api_call'
	from /Users/blowmage/.gem/repos/gcloud-ruby/gems/google-gax-0.8.4/lib/google/gax/api_callable.rb:250:in `block in create_api_call'
	from /Users/blowmage/google/codez/gcloud-ruby/google-cloud-pubsub/lib/google/cloud/pubsub/v1/publisher_client.rb:326:in `publish'
	from benchmark_gapic.rb:56:in `block (4 levels) in <main>'
```

Here is the script:

```ruby
$stdout.sync = true
Thread.abort_on_exception = true

require "google/cloud/pubsub/v1"
require "grpc"
require "googleauth"
require "securerandom"
require "benchmark"

# This will benchmark publishing 1 million 4KB messages
# using the previous batch method, and the new async publisher

project = ENV["GOOGLE_CLOUD_PROJECT"] || "<ENTER-PROJECT-ID-HERE>"
keyfile = ENV["GOOGLE_CLOUD_KEYFILE"] || "<ENTER-KEYFILE-PATH-HERE>"

host = Google::Cloud::Pubsub::V1::PublisherClient::SERVICE_ADDRESS
chan_args = { "grpc.max_send_message_length"    => -1,
              "grpc.max_receive_message_length" => -1 }
auth = Google::Auth::ServiceAccountCredentials.make_creds \
  json_key_io: File.open(keyfile, mode: "r"),
  scope: Google::Cloud::Pubsub::V1::PublisherClient::ALL_SCOPES
chan_creds = GRPC::Core::ChannelCredentials.new.compose \
  GRPC::Core::CallCredentials.new(auth.updater_proc)
channel = GRPC::Core::Channel.new host, chan_args, chan_creds
pub_client = Google::Cloud::Pubsub::V1::PublisherClient.new channel: channel

topic_name = "loadtest"
topic_path = Google::Cloud::Pubsub::V1::PublisherClient.topic_path project, topic_name
topic_grpc = begin
               pub_client.get_topic topic_path
             rescue Google::Gax::GaxError => e
               raise e unless e.cause.code == 5
               pub_client.create_topic topic_path
             end

msg_data = SecureRandom.random_bytes 4*1024
messages = 1000.times.map { Google::Pubsub::V1::PubsubMessage.new data: msg_data }

n = (ARGV.first || 4).to_i

Benchmark.bm do |bm|
  bm.report "sequential" do
    # runs the publish requests synchronously
    responses = n.times.map do
      pub_response = pub_client.publish topic_path, messages
      puts pub_response # uncomment to see when recieved
      pub_response
    end
    # puts responses.inspect
  end

  bm.report "concurrent" do
    # runs the publish requests asynchronously
    threads = n.times.map do
      Thread.new do
        pub_response = pub_client.publish topic_path, messages
        puts pub_response # uncomment to see when recieved
        pub_response
      end
    end
    responses = threads.map &:value
    # puts responses.inspect
  end
end
```

This script is using the GAPIC layer to make the calls. I have a [version of this script](https://github.com/grpc/grpc/issues/11776#issuecomment-314892969) that uses the GRPC stub instead and is able to make more than 4 concurrent requests without locking up.

**Comment by geigerj** [comment]
@blowmage I'm not able to repro this on my workstation. I've tried values of 4, 6, 16, 32, and 100 without error. In your original stacktrace, the RetryError indicates cause `4:Deadline Exceeded`, which *could* be a server-side error. It could also be a client side associated with the retry logic, but I don't have a good way to tell without a repro.

If you're able to reproduce this problem consistently, could you run the n=6 case with environment variables `GRPC_TRACE=http GRPC_VERBOSITY=debug` and send me a gist of the output so I can get a sense of what the client is actually doing?

**Comment by blowmage** [comment]
> I'm not able to repro this on my workstation.

Are you using Linux or OS X? The the system I'm seeing this on is OS X. I haven't tried Linux.

>  In your original stacktrace, the RetryError indicates cause 4:Deadline Exceeded, which could be a server-side error. It could also be a client side bug associated with the retry logic, but I don't have a good way to tell without a repro.

The Deadline Exceeded is raised by GAX after the concurrent requests are locked up. 6 sequential requests take ~20 seconds to complete, but the 6 concurrent requests raise an error after ~10 minutes.

> send me a gist of the output so I can get a sense of what the client is actually doing?

All ~82,000 lines of output are here: https://gist.github.com/blowmage/7a8fee4d7a98ac5e0a70fe33d4299cc2

**Comment by blowmage** [comment]
You might also try to throttle your network connection to ensure that each request doesn't complete too quickly.

**Comment by geigerj** [comment]
Ah, should have clarified, I'm on Linux. As you suggested, it's possible that I'm not seeing the same behavior because of platform (Linux) or network differences. Will try again tomorrow on Mac/limited bandwidth.

From the gist I see:
* the cancellation is client-side, not server-side, as you suspected.
* The exponential backoff alogorithm is not resetting the deadline even after a successful call completion. This is a bug (Note that the first 12 lines are from the 6 sequential calls, which are successful, yet the deadline still increases after each.)
```
grep -on "deadline=[0-9]\{0,3\}\.[0-9]*" benchmark_gapic.txt
    [GetTopic call removed]
59:deadline=12.501181822
60:deadline=12.501181822
384:deadline=15.698871160
385:deadline=15.698871160
736:deadline=18.781788669
737:deadline=18.781788669
1086:deadline=21.873670598
1087:deadline=21.873670598
1428:deadline=25.009020244
1429:deadline=25.009020244
1779:deadline=28.172765777
1780:deadline=28.172765777
2136:deadline=31.298385359
2137:deadline=31.298385359
2216:deadline=31.315516183
2217:deadline=31.315516183
2231:deadline=31.320804637
2232:deadline=31.320804637
2245:deadline=31.327169627
2246:deadline=31.327169627
2259:deadline=31.333615190
2260:deadline=31.333615190
2273:deadline=31.369117103
2274:deadline=31.369117103
3791:deadline=43.778861497
3792:deadline=43.778861497
3853:deadline=43.792442324
3854:deadline=43.792442324
3868:deadline=43.836202191
3869:deadline=43.836202191
3882:deadline=43.855804367
3883:deadline=43.855804367
3896:deadline=43.850086390
3897:deadline=43.850086390
3910:deadline=43.862956959
3911:deadline=43.862956959
...
81922:deadline=619.320335916
```

Will take a closer look tomorrow. I am also curious whether
* The concurrent calls are succeeding, and GAX is retrying them anyway
* GAX is actually receiving no response from the server after 619s

**Comment by geigerj** [comment]
Updates:

* I misread the gRPC trace above; the numbers are a deadline, not a timeout.
* I reproduced the buggy behavior by throttling my network to 2Mbps
* I verified that every outgoing GAPIC publish call has a timeout of 12s. This is expected given the JSON configuration file for the GAPIC.
* When the client "locks up" it is actually stuck in a loop of trying to publish with a timeout of 12s, timing out, and retrying with the exact same timeout.
* I verified that the calls succeed when the GAPIC configuration file sets the timeout to 120s.

I am fairly certain that the fix is just to change the GAPIC config. I just need to track down what the right values *should* be.
